src/crypto/tls/handshake_client.go

```diff
@@ -119,23 +123,65 @@ func (c *Conn) makeClientHello(minVersion uint16) (*clientHelloMsg, ecdheParamet
 	}
 
 	var params ecdheParameters
+	var keyShares []keyShare
+	var keySharePrivates []clientKeySharePrivate
+
 	if hello.supportedVersions[0] == VersionTLS13 {
 		hello.cipherSuites = append(hello.cipherSuites, defaultCipherSuitesTLS13()...)
 
-		curveID := config.curvePreferences()[0]
-		if _, ok := curveForCurveID(curveID); curveID != X25519 && !ok {
-			return nil, nil, errors.New("tls: CurvePreferences includes unsupported curve")
-		}
-		params, err = generateECDHEParameters(config.rand(), curveID)
-		if err != nil {
-			return nil, nil, err
```

Comment:
Can the following code be reworked so that we don't have to write down this code twice? It would make it easier to read and maintain.
```go
              curveID := config.curvePreferences()[0]
              if _, ok := curveForCurveID(curveID); curveID != X25519 && !ok {
                  return nil, nil, errors.New("tls: CurvePreferences includes unsupported curve")
              }        
              params, err = generateECDHEParameters(config.rand(), curveID)
              if err != nil {
                  return nil, nil, err
              }        
              keyShares = append(keyShares, keyShare{group: curveID, data: params.PublicKey()})
              keySharePrivates = append(keySharePrivates, params)
``

Reply:
Sure. I'll try refactoring it.

Reply:
Leaving this for next next week.